discovery-infra/delete_nodes.py

```diff
@@ -12,63 +15,113 @@
-            "-var-file=terraform.tfvars.json" % consts.TF_FOLDER
-        )
-        utils.run_command_with_output(cmd)
-    except:
-        log.exception("Failed to run terraform delete, deleting %s", consts.TF_FOLDER)
-        shutil.rmtree(consts.TF_FOLDER)
-    finally:
-        virsh_cleanup.clean_virsh_resources(
-            virsh_cleanup.DEFAULT_SKIP_LIST,
-            [
-                tfvars.get("cluster_name", consts.TEST_INFRA),
-                tfvars.get("libvirt_network_name", consts.TEST_INFRA),
-            ],
+@utils.on_exception(message='Failed to delete cluster')
+def try_to_delete_cluster(namespace, tfvars):
+    """ Try to delete cluster if assisted-service is up and such cluster
+        exists.
+    """
+    cluster_id = tfvars.get('cluster_inventory_id')
+    if not cluster_id:
+        return
+
+    args.namespace = namespace
+    client = assisted_service_api.create_client(
+        url=utils.get_assisted_service_url_by_args(args)
+    )
+    client.delete_cluster(cluster_id=cluster_id)
+
+
+def delete_nodes(cluster_name, namespace, tf_folder, tfvars):
+    """ Runs terraform destroy and then cleans it with virsh cleanup to delete
+        everything relevant.
+    """
+    @utils.on_exception(
+        message='Failed to run terraform delete',
+        silent=True
+    )
+    def _try_to_delete_nodes():
+        log.info('Start running terraform delete')
+        utils.run_command_with_output(
+            f'cd {tf_folder} && '
+            'terraform destroy '
+            '-auto-approve '
+            '-input=false '
+            '-state=terraform.tfstate ' 
+            '-state-out=terraform.tfstate ' 
+            '-var-file=terraform.tfvars.json'
         )
 
+    _try_to_delete_nodes()
+
+    default_network_name = consts.NETWORK_NAME_PREFIX + namespace
+    _delete_virsh_resources(
+        tfvars.get('cluster_name', cluster_name),
+        tfvars.get('libvirt_network_name', default_network_name),
+    )
+
+    log.info('Deleting %s', tf_folder)
+    shutil.rmtree(tf_folder)
+
+
+def _delete_virsh_resources(*filters):
+    log.info('Deleting virsh resources (filters: %s)', filters)
+    virsh_cleanup.clean_virsh_resources(
+        skip_list=virsh_cleanup.DEFAULT_SKIP_LIST,
+        resource_filter=filters
+    )
+
 
-# Deletes every single virsh resource, leaves only defaults
-def delete_all():
-    log.info("Deleting all virsh resources")
-    virsh_cleanup.clean_virsh_resources(virsh_cleanup.DEFAULT_SKIP_LIST, None)
+@utils.on_exception(
+    message='Failed to delete clusters from namespaces',
+    silent=True
+)
+def delete_clusters_from_all_namespaces():
+    for dirname in os.listdir(consts.TF_FOLDER):
+        res = _get_name_and_namespace_from_dirname(dirname)
+        if not res:
+            continue
+        name, namespace = res
+        delete_cluster(name, namespace)
+
+
+def _get_name_and_namespace_from_dirname(dirname):
```

Comment:
lets have one function in utils or new common file that handles all things relevant to this parsing and folder creation